### PR TITLE
#18798 escape wild cards during primary and imported keys loading

### LINKED ISSUE #18798: Snowflake - getPrimaryKeys()
### Description

It appears that DBeaver calls the getPrimaryKeys() method without schema/table names which in some cases can cause performance problems with Snowflake (underlying Snowflake JDBC driver yields a 'show primary keys in database ...' command rather than a 'show primary keys in table xxxx.xxxx.xxxx'.

### DBeaver Version

Community Edition 22.3.2.202301060755

### Operating System

Win 10 (21H2)

### Database and driver

Snowflake JDBC - 3.13.25

### Steps to reproduce

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by LonwoLonwo**
2023-01-18 19_30_19-DBeaver 22 3 3 - _TEST_DB_ Script-36

Please check

- [x] primary keys loading (CE) for Snowflake for a table with "_" with driver >= 3.13.19
- [x] primary keys loading (CE) for Snowflake for a schema with "_" with driver >= 3.13.19
- [x] foreign keys loading for Snowflake for a table with "_"  with driver >= 3.13.19
- [x] foreign keys loading for Snowflake for a schema with "_" with driver >= 3.13.19
- [x] primary keys loading (CE) for Snowflake for a table with "_" with driver < 3.13.19
- [x] primary keys loading (CE) for Snowflake for a schema with "_" with driver < 3.13.19
- [x] foreign keys loading for Snowflake for a table with "_"  with driver < 3.13.19
- [x] foreign keys loading for Snowflake for a schema with "_" with driver < 3.13.19

### CODE REVIEW

**Review by ShadelessFox — APPROVED** [review]

**Review by E1izabeth — APPROVED** [review]

**Comment by Matvey16** [comment]
Keys are not shown for a table with "_"

dbeaver_DqdwnAKsG3

**Comment by LonwoLonwo** [comment]
Yeah, because they added support of wildcards to the primary keys reading method only in May 
https://github.com/snowflakedb/snowflake-jdbc/commit/10f576afafec3eb488e66051870f4df5f4aade1b
driver version ~ 3.13.19
I am thinking about our options

**Review by E1izabeth — APPROVED** [review]

File: plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeMetaModel.java [hunk 1]
```diff
@@ -137,4 +137,65 @@ public DBCQueryTransformer createQueryTransformer(@NotNull DBCQueryTransformType
...
+        boolean recognizeWildCards = isRecognizeWildCards(session, owner);
+        GenericSchema schema = owner.getSchema();
+        String schemaName = getSchemaNameForPattern(session, recognizeWildCards, schema);
+        String tableName = getTableNameForPattern(session, owner, forParent, recognizeWildCards);
+        return session.getMetaData()
+            .getImportedKeys(owner.getCatalog() == null ? null : owner.getCatalog().getName(), schemaName, tableName)
+            .getSourceStatement();
+    }
+
+    private boolean isRecognizeWildCards(@NotNull JDBCSession session, @NotNull GenericStructContainer owner) throws SQLException {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeMetaModel.java]: Maybe just `supportsWildcards`?

File: plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeMetaModel.java [hunk 2]
```diff
@@ -137,4 +137,65 @@ public DBCQueryTransformer createQueryTransformer(@NotNull DBCQueryTransformType
...
+        // Snowflake driver do not recognize wild cards patterns before version 3.13.19 - and 19 here is the number of patch, not minor
+        if (owner.getDataSource().isDriverVersionAtLeast(3, 13)) {
+            String driverVersion = session.getMetaData().getDriverVersion();
+            if (CommonUtils.isNotEmpty(driverVersion) && driverVersion.contains(".")) {
+                String[] strings = driverVersion.split("\\.");
+                if (strings.length == 3) {
+                    return CommonUtils.toLong(strings[2]) >= 19;
+                }
+            }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeMetaModel.java:180]: I know it's temporary, but maybe move this method to the data source and make its signature look like:
```java
boolean isDriverVersionAtLeast(int major, int minor, int patch)
```

?

Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeMetaModel.java:180]: I think it is overcoding. I will move this method if I have another place to use it.

File: plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeMetaModel.java (hunk 1 above)
Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeMetaModel.java]: I do not see any reason for this, but ok

**Review by ShadelessFox — APPROVED** [review]

**Comment by Matvey16** [comment]
Verified

**Pull request merged**